qiskit_optimization/applications/max_k_cut.py

```diff
@@ -0,0 +1,198 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+
+"""An application class for the Max-k-cut."""
+
+from typing import List, Dict, Tuple, Optional, Union
+import networkx as nx
+import numpy as np
+from docplex.mp.model import Model
+
+from qiskit_optimization.algorithms import OptimizationResult
+from qiskit_optimization.problems.quadratic_program import QuadraticProgram
+from qiskit_optimization.translators import from_docplex_mp
+from .graph_optimization_application import GraphOptimizationApplication
+
+try:
+    from matplotlib.pyplot import cm
+    from matplotlib.colors import to_rgba
+
+    _HAS_MATPLOTLIB = True
+except ImportError:
+    import random as rd
+
+    _HAS_MATPLOTLIB = False
+
+
+class Maxkcut(GraphOptimizationApplication):
+    """Optimization application for the "max-k-cut" [1] problem based on a NetworkX graph.
+
+    References:
+        [1]: Z. Tabi et al.,
+             "Quantum Optimization for the Graph Coloring Problem with Space-Efficient Embedding"
+             2020 IEEE International Conference on Quantum Computing and Engineering (QCE),
+             2020, pp. 56-62, doi: 10.1109/QCE49297.2020.00018.,
+             https://ieeexplore.ieee.org/document/9259934
+    """
+
+    def __init__(
+        self,
+        graph: Union[nx.Graph, np.ndarray, List],
+        k: int,
+        colors: Optional[Union[List[str], List[List[int]]]] = None,
+    ) -> None:
+        """
+        Args:
+            graph: A graph representing a problem. It can be specified directly as a
+                `NetworkX <https://networkx.org/>`_ graph,
+                or as an array or list format suitable to build out a NetworkX graph.
+            k: The number of colors
+            colors: List of strings or list of colors in rgba lists to be assigned to each
+                resulting subset, there must be as many colors as the number k
+        """
+        super().__init__(graph=graph)
+        self._subsets_num = k
+        self._colors = colors if colors and len(colors) >= k else None
```

Comment:
For colors I see that more are allowed to be defined and presumably ignored; however if colors were given but less than k then they are discarded. In looking further, from what can see, these are just used if/when the user the draws the result.

I am wondering whether it might be better to move this off the constructor and just say have a method call something like `set_draw_colors` that takes the same thing and stores it for the rendering. That makes this feel less like part of the problem definition - at least to me. I would have said pass them on draw but then this is implementing logic defined in the base class and it does not seem to support customizing the rendering in this way.

In terms of the colors I think it should define what the string and int formats are - if its simply matplot color format then I think that can be stated perhaps with a link to matplot colot doc that has the info plus perhaps an example in the code to help people.

Also I wonder since this is a set of colors should they be passed in as a set so that it contains unique items - otherwise a user can specify the same color again. In regards of the behavior or more or less colors given than k maybe you prefer the less severe mechanism of discarding all or part of them depending on size. I think at least the code should emit a warning to alert the user - ideally they should be passing in k colors right not some other sized set. If this is a separate method just to set draw colors then maybe allowing only None (to reset to internal default) or a k sized set and raising a ValueError if the size if not equal to k might be more acceptable?

Reply:
Thank you for your comments, they have already been addressed:

- The importance of including the definition of colors in the constructor is that this implementation is based on the "Graph Coloring Problem", but I understand your concern, we can have the solution of the max-k-cut problem without the need to draw the resulting graph. To change (or add) the colors you have the methods set/get (`colors`, called like that for the same reason, that we are thinking about the "Graph Coloring Problem").
So, I have removed the `colors` parameter from the constructor, (before it was optional), I have already done this update.

- As for the documentation, certainly including an example is much better for people who use this class, that update is ready, please review.

- Regarding that the user can specify a repeated color, it seems to me that we should allow this type of flexibility, that this decision is up to the user, finally if it was not his/her intention, she/he will be able to see the result when he requests the drawing of the graph (the user will see fewer colors than '`k`').

- Now it is being validated that the size of the color list is equal to the k parameter, and if it is not, the `ValueError` is thrown indicating that the colors have not been assigned because they are not the correct amount.

Reply:
@woodsp-ibm, what do you think of the changes?